bundle/edu.gemini.sp.vcs/src/main/scala/edu/gemini/sp/vcs/diff/StaffOnlyFieldCorrection.scala

```diff
@@ -0,0 +1,108 @@
+import edu.gemini.pot.sp.{ISPStaffOnlyFieldProtected, SPNodeKey}
+import edu.gemini.shared.util.VersionComparison.Newer
+import edu.gemini.sp.vcs.diff.MergeCorrection._
+import edu.gemini.spModel.core.SPProgramID
+import edu.gemini.spModel.data.ISPDataObject
+import edu.gemini.spModel.gemini.obscomp.SPProgram
+import edu.gemini.spModel.rich.pot.sp._
+
+import java.security.Permission
+
+import edu.gemini.util.security.permission.{VisitorPermission, StaffPermission}
+
+import scalaz._
+import Scalaz._
+
+/** Corrections required to reset any data object fields to which the user does
+  * not have permission.  Some data, for example time accounting information or
+  * private notes, may only be edited by staff members. If the user manages to
+  * edit these fields either via an OT bug or malicious editing of an exported
+  * program, they must be reset when synchronized with the remote database. */
+class StaffOnlyFieldCorrection(lifespanId: LifespanId, key: SPNodeKey, pid: SPProgramID, remote: SPNodeKey => Option[Modified]) extends CorrectionAction {
+
+  import StaffOnlyFieldCorrection._
+
+  private def resetFields(mp: MergePlan): MergePlan = {
+    def corrected(m: Modified, dob: StaffProtected, correct: StaffProtected => Unit): MergeNode =
+      m.copy(nv = m.nv.incr(lifespanId), dob = dob.copy <| correct)
+
+    def correctedNew(m: Modified, dob: StaffProtected): MergeNode =
+      if (dob.staffOnlyFieldsDefaulted()) m
+      else corrected(m, dob, _.resetStaffOnlyFieldsToDefaults())
+
+    def correctedExisting(m: Modified, dob: StaffProtected, existing: ISPDataObject): MergeNode =
+      if (dob.staffOnlyFieldsEqual(existing)) m
+      else corrected(m, dob, _.setStaffOnlyFieldsFrom(existing))
+
+    def correctedNode(mn: MergeNode): MergeNode =
+      mn match {
+        case m@Modified(k, nv, dob: StaffProtected, _) =>
+          remote(k).fold(correctedNew(m, dob)) { remoteMod =>
+            nv.compare(remoteMod.nv) match {
+              case Newer => correctedExisting(m, dob, remoteMod.dob)
+              case _     => mn
+            }
+          }
+
+        case _ => mn
+      }
+
+    mp.copy(update = mp.update.map(correctedNode))
+  }
+
+  def apply(mp: MergePlan, hasPermission: PermissionCheck): VcsAction[MergePlan] = {
+    // Regular staff permission is determined by program contact information. To
+    // prevent changing the contact and gaining staff permission we have to check
+    // that contacts aren't edited unless you have super-staff permission.
+    val contactsMatch = contact(mp.update.rootLabel) == remote(key).flatMap(contact)
+
+    canSetStaffFields(pid, contactsMatch, hasPermission).map { canSet =>
+      if (canSet) mp else resetFields(mp)
+    }
+  }
+}
+
+object StaffOnlyFieldCorrection {
+  type StaffProtected = ISPDataObject with ISPStaffOnlyFieldProtected
+
+  private def contact(mn: MergeNode): Option[String] = {
+    def contact(sp: SPProgram) = Option(sp.getContactPerson).map(_.trim.toLowerCase)
```

Comment:
Yes, it can return `null`.  It's part of the ancient `SPProgram` data object for the top-level node. It is initialized to the empty string but can be assigned `null`.  We pretty much have to assume that anything coming from the old Java model will sometimes be `null`.